pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,225 @@
+
+    def _cleanup(self):
+        # reset the param values to those stored before the hmc run
+        for name, grad in self._cached_param_grads.items():
+            param = pyro.get_param_store().get_param(name)
+            param.grad = grad
+
+    def _traces(self, *args, **kwargs):
+        self._setup(*args, **kwargs)
+        # sample p's from the distribution given by p_dist
+        r_dist = {}
+        z = {name: node['value'] for name, node in self._prototype_trace.iter_stochastic_nodes()}
+        for name, value in z.items():
+            r_mu = torch_zeros_like(value)
+            r_sigma = torch_ones_like(value)
+            r_dist[name] = dist.Normal(mu=r_mu, sigma=r_sigma)
+
+        # Run HMC iterations
+        t = 0
+        print('Simulating using HMC...')
+        while t < self.warmup_steps + self.num_samples:
+            if t % 100 == 0:
+                print('Iteration: {}'.format(t))
+            # sample momentum
+            r = {name: pyro.sample('r_{}_t={}'.format(name, t), r_dist[name]) for name in z}
+            z_new, r_new = verlet_integrator(z,
+                                             r,
+                                             self._grad_potential,
+                                             self.step_size,
+                                             self.num_steps)
+            energy_proposal = self._energy(z_new, r_new)
+            energy_current = self._energy(z, r)
+            # print("Energy - current: {}".format(energy_current))
+            # print("Energy - proposal: {}".format(energy_proposal))
+            delta_energy = energy_proposal - energy_current
+            t += 1
+            rand = pyro.sample('rand_t='.format(t), dist.uniform, a=ng_zeros(1), b=ng_ones(1))
+            if rand.log().data[0] < -delta_energy.data[0]:
+                self._accept_cnt += 1
+                z = z_new
+            if t < self.warmup_steps:
+                continue
+            yield (z, self._log_prob(z))
+        self._cleanup()
+
+    @property
+    def acceptance_ratio(self):
+        if self._accept_cnt is None:
+            raise AttributeError("The attribute _accept_cnt is not updated until the first simulation run.")
+        return self._accept_cnt / (self.warmup_steps + self.num_samples)
+
+
+def test_normal_normal():
+    def model(data):
+        mu = pyro.param('mu', Variable(torch.zeros(10), requires_grad=True))
+        x = pyro.sample('x', dist.normal, mu=mu, sigma=Variable(torch.ones(10)))
+        pyro.sample('data', dist.normal, obs=data, mu=x, sigma=Variable(torch.ones(10)))
+
+    data = Variable(torch.ones(1, 10))
+    hmc = HMC(model, step_size=0.6, num_steps=3, num_samples=400, warmup_steps=50)
+    traces = []
+    for t, _ in hmc._traces(data):
+        traces.append(t['x'])
+    print('Acceptance ratio: {}'.format(hmc.acceptance_ratio))
+    print('Posterior mean:')
+    print(torch.mean(torch.stack(traces), 0).data)
```

Comment:
It would be a good end-to-end test of the math to check the true-vs-empirical mean and standard deviation.

Reply:
Yes, will add tests for that. This was only meant for my debugging. :)